flang/lib/Lower/Bridge.cpp

```diff
@@ -351,36 +338,72 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+
+      mlir::Type symType = getSymbolAddress(sym).getType();
+      auto refTy = symType.dyn_cast<fir::ReferenceType>();
+      symType = refTy.getEleTy();
+
+      auto allocaVal = allocateThreadLocal(
+          sym, symType, fir::factory::getExtents(*builder, loc, cexv),
+          fir::getTypeParams(cexv));
+
+      for (const auto &mem : commonDet->objects()) {
+        const auto *memDet =
+            mem->detailsIf<Fortran::semantics::ObjectEntityDetails>();
+        assert(memDet != nullptr && "No common member object entity found");
+        Fortran::lower::SymbolBox memCsb = lookupSymbol(*mem);
+        auto *cvtSymOp = memCsb.getAddr().getDefiningOp();
+        auto *coordOp = cvtSymOp->getOperand(0).getDefiningOp();
+        auto cvtBlkVal = builder->createConvert(
+            loc, coordOp->getOperand(0).getType(), allocaVal);
+        auto coordVal = builder->create<fir::CoordinateOp>(
+            loc, cvtSymOp->getOperand(0).getType(), cvtBlkVal,
+            coordOp->getOperand(1));
+        auto cvtSymVal =
+            builder->createConvert(loc, memCsb.getAddr().getType(), coordVal);
+
+        fir::ExtendedValue memCexv = getExtendedValue(memCsb);
+        auto memExval = memCexv.match(
+            [&](const fir::CharBoxValue &box) -> fir::ExtendedValue {
+              return fir::CharBoxValue(
+                  cvtSymVal, fir::factory::readCharLen(*builder, loc, box));
+            },
+            [&](const fir::ArrayBoxValue &box) -> fir::ExtendedValue {
+              return fir::ArrayBoxValue{
+                  cvtSymVal, fir::factory::getExtents(*builder, loc, box)};
+            },
+            [&](const fir::CharArrayBoxValue &box) -> fir::ExtendedValue {
+              return fir::CharArrayBoxValue{
+                  cvtSymVal, fir::factory::readCharLen(*builder, loc, box),
+                  fir::factory::getExtents(*builder, loc, box)};
+            },
+            [&](const fir::MutableBoxValue &box) -> fir::ExtendedValue {
+              TODO(loc, "Pointer in common block not implemented");
+            },
+            [&](const auto &) -> fir::ExtendedValue { return cvtSymVal; });
+        localSymbols.addSymbol(*mem, memExval, true);
+      }
+    } else if (const auto *hostAssocDet =
+                   sym.detailsIf<Fortran::semantics::HostAssocDetails>()) {
+      assert(hostAssocDet != nullptr &&
+             "No common block or host-association found");
+      const Fortran::semantics::Symbol &hsym = hostAssocDet->symbol();
+      Fortran::lower::SymbolBox hsb = lookupSymbol(hsym);
+      fir::ExtendedValue hexv = getExtendedValue(hsb);
+      auto allocaVal = allocateThreadLocal(
+          sym, genType(sym), fir::factory::getExtents(*builder, loc, hexv),
+          fir::getTypeParams(hexv));
+      auto exval = fir::substBase(hexv, allocaVal);
+      if (!bindSymbol(sym, exval))
+        return false;
+    } else {
+      return false;
+    }
+    return true;
```

Comment:
These `return`s are a bit of an eye-snag. https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return